src/Microsoft.AspNet.Http/Features/FormFeature.cs

```diff
@@ -129,8 +153,12 @@ public IFormCollection ReadForm()
 
                             var file = new FormFile(_request.Body, section.BaseStreamOffset.Value, section.Body.Length)
                             {
-                                Headers = headers,
+                                Headers = new HeaderDictionary(section.Headers),
```

Comment:
Another optimization might be to use a different collection type that implement IHeaderDictionary. I'm assuming this sub header dictionary normally has a tiny number of items in the most common cases. Something to consider in the future.

/cc @Tratcher